src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -0,0 +1,794 @@
+ * variadic function by using a <em>specialized</em> method type and function descriptor: for each argument that is to be
+ * passed as a variadic argument, an explicit, additional, carrier type and memory layout must be present in the method type and
+ * function descriptor objects passed to the linker. Furthermore, as memory layouts corresponding to variadic arguments in
+ * a function descriptor must contain additional classification information, it is required that
+ * {@link #asVarArg(MemoryLayout)} is used to create the memory layouts for each parameter corresponding to a variadic
+ * argument in a specialized function descriptor.
+ *
+ * @apiNote In the future, if the Java language permits, {@link CLinker}
+ * may become a {@code sealed} interface, which would prohibit subclassing except by
+ * explicitly permitted types.
+ *
+ * @implSpec
+ * Implementations of this interface are immutable, thread-safe and <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>.
+ */
+public interface CLinker {
+
+    /**
+     * Returns the C linker for the current platform.
+     * <p>
+     * This method is <em>restricted</em>. Restricted method are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
+     * @return a linker for this system.
+     * @throws IllegalAccessError if the runtime property {@code foreign.restricted} is not set to either
+     * {@code permit}, {@code warn} or {@code debug} (the default value is set to {@code deny}).
+     */
+    static CLinker getInstance() {
+        Utils.checkRestrictedAccess("CLinker.getInstance");
+        return SharedUtils.getSystemLinker();
+    }
+
+    /**
+     * Obtain a foreign method handle, with given type, which can be used to call a
+     * target foreign function at a given address and featuring a given function descriptor.
+     *
+     * @param symbol   downcall symbol.
+     * @param type     the method type.
+     * @param function the function descriptor.
+     * @return the downcall method handle.
+     * @throws IllegalArgumentException in the case of a carrier type and memory layout mismatch.
+     */
+    MethodHandle downcallHandle(Addressable symbol, MethodType type, FunctionDescriptor function);
+
+    /**
+     * Allocates a native segment whose base address (see {@link MemorySegment#address}) can be
+     * passed to other foreign functions (as a function pointer); calling such a function pointer
+     * from native code will result in the execution of the provided method handle.
+     *
+     * <p>The returned segment is <em>not</em> thread-confined, and it only features
+     * the {@link MemorySegment#CLOSE} access mode. When the returned segment is closed,
```

Comment:
`The returned segment is <em>not</em> thread-confined` ? Since it features CLOSE, it can be closed explicitly - I'm not sure 100% of what additional clarification is required - but I'm happy to make this clearer (I need more info).